numba/listobject.py

```diff
@@ -0,0 +1,1062 @@
+            front += 1
+            back -= 1
+
+    return impl
+
+
+@overload_method(types.ListType, 'copy')
+def impl_copy(l):
+    if isinstance(l, types.ListType):
+        def impl(l):
+            return l[:]
+
+        return impl
+
+
+@overload_method(types.ListType, 'index')
+def impl_index(l, item, start=None, end=None):
+    if not isinstance(l, types.ListType):
+        return
+    itemty = l.item_type
+
+    def check_arg(arg, name):
+        if not (arg is None
+                or arg in types.signed_domain
+                or isinstance(arg, (types.Omitted, types.NoneType))):
+            raise TypingError("{} argument for index must be a signed integer"
+                              .format(name))
+    check_arg(start, "start")
+    check_arg(end, "end")
+
+    def impl(l, item, start=None, end=None):
+        casteditem = _cast(item, itemty)
+        for i in handle_slice(l, slice(start, end, 1)):
+            if l[i] == casteditem:
+                return i
+        else:
+            raise ValueError("item not in list")
+
+    return impl
+
+
+@lower_builtin('getiter', types.ListType)
+def impl_list_getiter(context, builder, sig, args):
+    """Implement iter(List).
+    """
+    [tl] = sig.args
+    [l] = args
+    iterablety = types.ListTypeIterableType(tl)
+    it = context.make_helper(builder, iterablety.iterator_type)
+
+    fnty = ir.FunctionType(
+        ir.VoidType(),
+        [ll_listiter_type, ll_list_type],
+    )
+
+    fn = builder.module.get_or_insert_function(fnty, name='numba_list_iter')
+
+    proto = ctypes.CFUNCTYPE(ctypes.c_size_t)
+    listiter_sizeof = proto(_helperlib.c_helpers['list_iter_sizeof'])
+    state_type = ir.ArrayType(ir.IntType(8), listiter_sizeof())
+
+    pstate = cgutils.alloca_once(builder, state_type, zfill=True)
+    it.state = _as_bytes(builder, pstate)
+    it.parent = l
+
+    dp = _container_get_data(context, builder, iterablety.parent, args[0])
+    builder.call(fn, [it.state, dp])
+    return impl_ret_borrowed(
+        context,
+        builder,
+        sig.return_type,
+        it._getvalue(),
+    )
+
+
+@overload(operator.eq)
+def impl_equal(this, other):
+    if not isinstance(this, types.ListType):
+        return
+    if not isinstance(other, types.ListType):
+        # If RHS is not a list, always returns False
+        def impl_type_mismatch(this, other):
+            return False
+        return impl_type_mismatch
+
+    otheritemty = other.item_type
+
+    # special case: either list has array types as items
+    if (isinstance(this.item_type, types.Array) or
+            isinstance(other.item_type, types.Array)):
+        def impl_type_matched_array_item_type(this, other):
+            if len(this) != len(other):
+                return False
+            for i in range(len(this)):
+                this_item, other_item = this[i], other[i]
+                # Cast item from LHS to the key-type of RHS
+                this_item = _cast(this_item, otheritemty)
```

Comment:
This cast is unnecessary.

Reply:
fixed in 86a1963d13458dfe59ef6d935fa30a643cfe69a0